File: sdk/python/ailink/resources/credentials.py
"""Resource for managing encrypted credentials."""

from typing import List, Dict, Any, Optional
from ..types import Credential
from ..exceptions import raise_for_status


class CredentialsResource:
    """Management API resource for credentials (metadata only — no secrets exposed)."""

    def __init__(self, client):
        self._client = client

    def list(self, project_id: Optional[str] = None) -> List[Credential]:
        """List credential metadata for a project."""
        params = {}
        if project_id:
            params["project_id"] = project_id
        resp = self._client._http.get("/api/v1/credentials", params=params)
        raise_for_status(resp)
        return [Credential(**item) for item in resp.json()]

    def create(
        self,
        name: str,
        provider: str,
        secret: str,
        project_id: Optional[str] = None,
        injection_mode: Optional[str] = None,
        injection_header: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Create a new encrypted credential.

        Returns a dict with the credential ``id`` and metadata.
        The secret is encrypted at rest and never returned.
        """
        payload: Dict[str, Any] = {
            "name": name,
            "provider": provider,
            "secret": secret,
        }
        if project_id:
            payload["project_id"] = project_id
        if injection_mode:
            payload["injection_mode"] = injection_mode
        if injection_header:
            payload["injection_header"] = injection_header
        resp = self._client._http.post("/api/v1/credentials", json=payload)
        raise_for_status(resp)
        return resp.json()


class AsyncCredentialsResource:
    """Async Management API resource for credentials."""

    def __init__(self, client):
        self._client = client

    async def list(self, project_id: Optional[str] = None) -> List[Credential]:
        """List credential metadata for a project."""
        params = {}
        if project_id:
            params["project_id"] = project_id
        resp = await self._client._http.get("/api/v1/credentials", params=params)
        raise_for_status(resp)
        return [Credential(**item) for item in resp.json()]

    async def create(
        self,
        name: str,
        provider: str,
        secret: str,
        project_id: Optional[str] = None,
        injection_mode: Optional[str] = None,
        injection_header: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Create a new encrypted credential."""
        payload: Dict[str, Any] = {
            "name": name,
            "provider": provider,
            "secret": secret,
        }
        if project_id:
            payload["project_id"] = project_id
        if injection_mode:
            payload["injection_mode"] = injection_mode
        if injection_header:
            payload["injection_header"] = injection_header
        resp = await self._client._http.post("/api/v1/credentials", json=payload)
        raise_for_status(resp)
        return resp.json()

File: sdk/typescript/src/resources/credentials.ts
import type { HttpClient } from "../http.js";
import type { Credential, CredentialCreateResponse, PaginationOptions, JsonObject } from "../types.js";

export class CredentialsResource {
    constructor(private readonly http: HttpClient) { }

    /** Store a new encrypted credential. */
    async create(options: { name: string; provider: string; secret: string; projectId?: string; injectionMode?: string; injectionHeader?: string }): Promise<CredentialCreateResponse> {
        const body: Record<string, unknown> = { name: options.name, provider: options.provider, secret: options.secret };
        if (options.projectId) body["project_id"] = options.projectId;
        if (options.injectionMode) body["injection_mode"] = options.injectionMode;
        if (options.injectionHeader) body["injection_header"] = options.injectionHeader;
        const res = await this.http.post("/api/v1/credentials", body);
        return (await res.json()) as CredentialCreateResponse;
    }

    /** List all credentials (keys are never returned). */
    async list(options: PaginationOptions = {}): Promise<Credential[]> {
        const res = await this.http.get("/api/v1/credentials", { params: { limit: options.limit, offset: options.offset } });
        return (await res.json()) as Credential[];
    }

    /** Get a credential by ID. */
    async get(credentialId: string): Promise<Credential> {
        const res = await this.http.get(`/api/v1/credentials/${credentialId}`);
        return (await res.json()) as Credential;
    }

    /** Delete a credential. */
    async delete(credentialId: string): Promise<JsonObject> {
        const res = await this.http.delete(`/api/v1/credentials/${credentialId}`);
        return (await res.json()) as JsonObject;
    }

    /** Rotate a credential — generates a new encryption, returns new metadata. */
    async rotate(credentialId: string, newApiKey: string): Promise<JsonObject> {
        const res = await this.http.post(`/api/v1/credentials/${credentialId}/rotate`, { api_key: newApiKey });
        return (await res.json()) as JsonObject;
    }
}
